### ISSUE TITLE
bug: Using DLQ for DLQ, polling/receiving from the first DLQ causes the message to go into the second DLQ immediately

### ISSUE THREAD

**Issue by eyalrin** [localstack/localstack#5225]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Defined 3 SQS queues:

1. Queue1 (Set it's DLQ to Queue2 using `REDRIVE_POLICY` with `maxReceiveCount=1` and `VISIBILITY_TIMEOUT=3`)
2. Queue2 (Set it's DLQ to Queue3 using `REDRIVE_POLICY`with `maxReceiveCount=1` and `VISIBILITY_TIMEOUT=3`)
3. Queue3 

Therefore the messages flow is `Queue1 -> Queue2 -> Queue3` (in case of no messages deletion and more then 1 receive operation on the queue).

**Flow:**
1. Send message to `Queue1` (`aws sqs send-message`)
2. Poll message from `Queue1` (`aws sqs receive-message`) and got message
3. Wait for `VISIBILITY_TIMEOUT` to expire
4. Poll message from `Queue1` (`aws sqs receive-message`) and got **no** message -> message goes to `Queue2`
5. Poll message from `Queue2` (`aws sqs receive-message`) - got **no** message -> message goes to `Queue3`

### Expected Behavior

Flow (Same queues configuration):
1. Send message to `Queue1` (`aws sqs send-message`)
2. Poll message from `Queue1` (`aws sqs receive-message`) and got message
3. Wait for `VISIBILITY_TIMEOUT` to expire
4. Poll message from `Queue1` (`aws sqs receive-message`) and got **no** message -> message goes to `Queue2`
5. Poll message from `Queue2` (`aws sqs receive-message`) and got message
6. Poll message from `Queue2` (`aws sqs receive-message`) - got **no** message -> message goes to `Queue3`

Look at `5` - This step does not happen actually in Localstack (tested same configuration on AWS and it works as expected)

### How are you starting LocalStack?

With a docker-compose file

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

    docker compose up -d

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

    awslocal sqs create-queue --queue-name queue3 --attributes file:///docker-entrypoint-initaws.d/sqs-config/queue3-attributes.json
    
    awslocal sqs create-queue --queue-name queue2 --attributes file:///docker-entrypoint-initaws.d/sqs-config/queue2-attributes.json
     
    awslocal sqs create-queue --queue-name queue1 --attributes file:///docker-entrypoint-initaws.d/sqs-config/queue1-attributes.json

(Remove the `.txt` from each file - added it just to be able to upload)
[queue1-attributes.json.txt](https://github.com/localstack/localstack/files/7812796/queue1-attributes.json.txt)
[queue2-attributes.json.txt](https://github.com/localstack/localstack/files/7812797/queue2-attributes.json.txt)
[queue3-attributes.json.txt](https://github.com/localstack/localstack/files/7812798/queue3-attributes.json.txt)

### Environment

```markdown
- OS: MacOS
- LocalStack: localstack-full:0.13.0.10
```

**Comment by whummer** [comment]
Hi @eyalrin , thanks for reporting. Indeed, there's currently a few issues with the default SQS provider.

We have recently added a new, improved SQS implementation, which can be selected by setting this config option:
```
PROVIDER_OVERRIDE_SQS=asf
```

Seems like with this option, your use case should be supported. Can you please give that a try and let us know? Thanks! (We'll look into making this the default SQS provider in the near future..)

**Comment by eyalrin** [comment]
Hi @whummer,

I Tried setting `.withEnv("PROVIDER_OVERRIDE_SQS", "asf")` in my code but it causes Localstack’s SQS to not load properly (I guess) because now I can’t create a queue…

**Comment by thrau** [comment]
hey @eyalrin, i see you are using 0.13.0.1, but the asf SQS provider was added in 0.13.3, would it be possible to upgrade localstack to that version?

**Comment by eyalrin** [comment]
Hi @thrau & @whummer,

I tried now with `localstack:0.13.3`, seems SQS loads properly now but when I create in code the queue and tries to get its URL like so:
```
sqsClient.createQueue(CreateQueueRequest.builder().queueName("queueName").build()).queueUrl()
````
I get: 
```
http://localhost:4566/000000000000/queueName
``` 
where before the change I used to get 
```
http://aws-localstack:55416/000000000000/queueName
```
Why is that? Seems like it doesn't respect `testcontainers` rules...I don't want to change all my settings now to a new URL structure...

**Comment by whummer** [comment]
Hi @eyalrin , thanks for the update. Looks like the new SQS provider didn't respect the `HOSTNAME_EXTERNAL` configuration. This should now be fixed via #5238 - can you please pull the `latest` Docker image one more time and give it a try? Thanks!

**Comment by eyalrin** [comment]
Hi @whummer,

Now I getting a mixture of the two:
```
http://aws-localstack:4566/000000000000/webhook-delivery-deadletter
```

i.e., `aws-localstack` instead of `localhost` but still `4566` instead of the `textcontainer port` (5 digits, start with 55)

**Comment by thrau** [comment]
hi @eyalrin, thanks for the detailed information. based on the requirements we have now, we've decided to make the following changes:

1. Always render Queue URLs, never create and store them
    1. always use the host that was called (context.request.host → which is the Host header field including port and scheme). this means clients always see the host part of the queue url as they called localstack.
    2. no longer rely on HOSTNAME_EXTERNAL, or other host configurations. Queue URLs are always resolved dynamically based on what host the client calls when calling the edge proxy.
2. with the context.request object which is an instance of HttpRequest (werkzeug)
3. remove name and URL-based indexes in SqsProvider → always generate QueueKeys from context

@baermat will make the changes and we think this will resolve your issues!

thanks for you patience, we'll get it over the line the next few days!

**Comment by eyalrin** [comment]
Hi @thrau, glad the information I provided helped you guys.
Once you finish your changes let me know and I'll re-check the scenario in question.
10X

**Comment by baermat** [comment]
HI @eyalrin , the changes are merged, please let us know if this works again for you, thanks!

**Comment by mvallim** [comment]
> ### Is there an existing issue for this?
> * [x]  I have searched the existing issues
> 
> ### Current Behavior
> Defined 3 SQS queues:
> 
> 1. Queue1 (Set it's DLQ to Queue2 using `REDRIVE_POLICY` with `maxReceiveCount=1` and `VISIBILITY_TIMEOUT=3`)
> 2. Queue2 (Set it's DLQ to Queue3 using `REDRIVE_POLICY`with `maxReceiveCount=1` and `VISIBILITY_TIMEOUT=3`)
> 3. Queue3
> 
> Therefore the messages flow is `Queue1 -> Queue2 -> Queue3` (in case of no messages deletion and more then 1 receive operation on the queue).
> 
> **Flow:**
> 
> 1. Send message to `Queue1` (`aws sqs send-message`)
> 2. Poll message from `Queue1` (`aws sqs receive-message`) and got message
> 3. Wait for `VISIBILITY_TIMEOUT` to expire
> 4. Poll message from `Queue1` (`aws sqs receive-message`) and got **no** message -> message goes to `Queue2`
> 5. Poll message from `Queue2` (`aws sqs receive-message`) - got **no** message -> message goes to `Queue3`
> 
> ### Expected Behavior
> Flow (Same queues configuration):
> 
> 1. Send message to `Queue1` (`aws sqs send-message`)
> 2. Poll message from `Queue1` (`aws sqs receive-message`) and got message
> 3. Wait for `VISIBILITY_TIMEOUT` to expire
> 4. Poll message from `Queue1` (`aws sqs receive-message`) and got **no** message -> message goes to `Queue2`
> 5. Poll message from `Queue2` (`aws sqs receive-message`) and got message
> 6. Poll message from `Queue2` (`aws sqs receive-message`) - got **no** message -> message goes to `Queue3`
> 
> Look at `5` - This step does not happen actually in Localstack (tested same configuration on AWS and it works as expected)
> 
> ### How are you starting LocalStack?
> With a docker-compose file
> 
> ### Steps To Reproduce
> #### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)
> ```
> docker compose up -d
> ```
> 
> #### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)
> ```
> awslocal sqs create-queue --queue-name queue3 --attributes file:///docker-entrypoint-initaws.d/sqs-config/queue3-attributes.json
> 
> awslocal sqs create-queue --queue-name queue2 --attributes file:///docker-entrypoint-initaws.d/sqs-config/queue2-attributes.json
>  
> awslocal sqs create-queue --queue-name queue1 --attributes file:///docker-entrypoint-initaws.d/sqs-config/queue1-attributes.json
> ```
> 
> (Remove the `.txt` from each file - added it just to be able to upload) [queue1-attributes.json.txt](https://github.com/localstack/localstack/files/7812796/queue1-attributes.json.txt) [queue2-attributes.json.txt](https://github.com/localstack/localstack/files/7812797/queue2-attributes.json.txt) [queue3-attributes.json.txt](https://github.com/localstack/localstack/files/7812798/queue3-attributes.json.txt)
> 
> ### Environment
> ```gfm
> - OS: MacOS
> - LocalStack: localstack-full:0.13.0.10
> ```
> 
> ### Anything else?
> _No response_

Hi @eyalrin,

The central problem here is that when a message leaves your main SQS and is moved to the SQS DLQ, the `receive_times` number is not reset to zero.

The ideal here is that whenever a message leaves a main SQS and is moved to another SQS DLQ that this behavior repeats itself.

This can be seen here:

https://github.com/localstack/localstack/blob/1b31f961d39b80602dbd96bf153426e7a98bb395/localstack/services/sqs/provider.py#L1070-L1071

Message is forwarded without resetting `receive_times`

The central idea:

```python
dead_message = std_m.message
dead_message.receive_times = 0
dl_queue.put(message=dead_message)
```

It's apparently a simple fix to make.

But it needs a better analysis, of possible impacts.

Thanks,

**Comment by thrau** [comment]
I took some time to dive into the issue once again, and can confirm that this is definitely working with our current provider. We do reset the receive time when we create a new `SqsMessage` object in the `put` operation here: https://github.com/localstack/localstack/blob/1b31f961d39b80602dbd96bf153426e7a98bb395/localstack/services/sqs/provider.py#L454
what we're passing to `dl_queue.put` is the original AWS SQS Message.

I did find a limitation related to the `ApproximateReceiveCount`, which is now also reset. I wrote a test against AWS demonstrating that it actually retains the `ApproximateReceiveCount` from previous receives even as it moves through DLQs. I tried implementing this behavior but gave up after an hour because it would involve either silly hacks or significant restructuring. I concluded it's not worth the effort right now, and documented the limitation here: https://github.com/localstack/docs/pull/223

So I'm closing this issue as resolved! Thanks @baermat for the initial implementation and @whummer for the test!
